File: aos_go/spawn.go
package main

import "math"

type RoutePoint struct {
	LatDeg    float64
	LonDeg    float64
	AltM      float64
	SpeedsKph float64
	Ecef      Ecef
}

type ObjectState struct {
	ID               string
	TeamID           string
	Role             Role
	StartSec         int
	Route            []RoutePoint
	SegmentEndSecs   []float64
	TotalDurationSec float64
	Position         Ecef
	DetectState      map[string]DetectionInfo
	HasDetonated     bool
}

func buildObjects(scenario *Scenario) []ObjectState {
	objects := make([]ObjectState, 0)

	// チームとオブジェクトを走査し、AoSの状態配列を組み立てます。
	for _, team := range scenario.Teams {
		for _, obj := range team.Objects {
			// 経路の緯度経度をECEF座標に変換して保持します。
			route := buildRoute(obj.Route)
			// 区間ごとの移動時間を計算し、後で位置補間に使います。
			segmentEnds, totalDuration := buildSegmentTimes(route)
			position := Ecef{}
			if len(route) > 0 {
				position = route[0].Ecef
			}

			objects = append(objects, ObjectState{
				ID:               obj.ID,
				TeamID:           team.ID,
				Role:             obj.Role,
				StartSec:         obj.StartSec,
				Route:            route,
				SegmentEndSecs:   segmentEnds,
				TotalDurationSec: totalDuration,
				Position:         position,
				DetectState:      make(map[string]DetectionInfo),
				HasDetonated:     false,
			})
		}
	}

	return objects
}

func buildRoute(route []Waypoint) []RoutePoint {
	result := make([]RoutePoint, 0, len(route))
	for _, wp := range route {
		// WGS84の緯度経度高度をECEF座標へ変換します。
		ecef := geodeticToECEF(wp.LatDeg, wp.LonDeg, wp.AltM)
		result = append(result, RoutePoint{
			LatDeg:    wp.LatDeg,
			LonDeg:    wp.LonDeg,
			AltM:      wp.AltM,
			SpeedsKph: wp.SpeedsKph,
			Ecef:      ecef,
		})
	}
	return result
}

func buildSegmentTimes(route []RoutePoint) ([]float64, float64) {
	if len(route) < 2 {
		return nil, 0
	}

	segmentEnds := make([]float64, 0, len(route)-1)
	acc := 0.0

	for i := 0; i < len(route)-1; i++ {
		// 区間距離と速度から移動に必要な秒数を算出します。
		a := route[i].Ecef
		b := route[i+1].Ecef
		distance := distanceECEF(a, b)
		speedMps := (route[i].SpeedsKph * 1000.0) / 3600.0
		duration := math.Inf(1)
		if speedMps > 0 {
			duration = distance / speedMps
		}
		acc += duration
		segmentEnds = append(segmentEnds, acc)
	}

	return segmentEnds, acc
}

File: aos_rs/src/spawn.rs
use crate::geo::geodetic_to_ecef;
use crate::geo::Ecef;
use crate::scenario::Scenario;
use crate::scenario::Waypoint;
use crate::sim::{ObjectState, RoutePoint};
use std::collections::HashMap;

pub fn build_objects(scenario: &Scenario) -> Vec<ObjectState> {
    let mut objects = Vec::new();

    // チームとオブジェクトを走査し、AoSの状態配列を組み立てます。
    for team in &scenario.teams {
        for obj in &team.objects {
            // 経路の緯度経度をECEF座標に変換して保持します。
            let route = build_route(&obj.route);
            // 区間ごとの移動時間を計算し、後で位置補間に使います。
            let (segment_end_secs, total_duration_sec) = build_segment_times(&route);
            let position_ecef = route.first().map(|p| p.ecef).unwrap_or(Ecef {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            });

            objects.push(ObjectState {
                id: obj.id.clone(),
                team_id: team.id.clone(),
                role: obj.role,
                start_sec: obj.start_sec,
                route,
                segment_end_secs,
                total_duration_sec,
                position_ecef,
                detect_state: HashMap::new(),
                has_detonated: false,
            });
        }
    }

    objects
}

fn build_route(route: &[Waypoint]) -> Vec<RoutePoint> {
    route
        .iter()
        .map(|wp| {
            // WGS84の緯度経度高度をECEF座標へ変換します。
            let ecef = geodetic_to_ecef(wp.lat_deg, wp.lon_deg, wp.alt_m);
            RoutePoint {
                lat_deg: wp.lat_deg,
                lon_deg: wp.lon_deg,
                alt_m: wp.alt_m,
                speeds_kph: wp.speeds_kph,
                ecef,
            }
        })
        .collect()
}

fn build_segment_times(route: &[RoutePoint]) -> (Vec<f64>, f64) {
    if route.len() < 2 {
        return (Vec::new(), 0.0);
    }

    let mut segment_end_secs = Vec::with_capacity(route.len() - 1);
    let mut acc = 0.0;

    for i in 0..route.len() - 1 {
        // 区間距離と速度から移動に必要な秒数を算出します。
        let a = route[i].ecef;
        let b = route[i + 1].ecef;
        let distance_m = crate::geo::distance_ecef(a, b);
        let speed_mps = (route[i].speeds_kph * 1000.0) / 3600.0;
        let duration = if speed_mps <= 0.0 {
            f64::INFINITY
        } else {
            distance_m / speed_mps
        };
        acc += duration;
        segment_end_secs.push(acc);
    }

    (segment_end_secs, acc)
}
